Various tailcall test improvements

jakobbotsch opened the pull request:
Add tests that test
* Tail calls to functions with small return values
* Tail calls to functions with retbuf
* Tail calls to functions with multi-reg returned structs
* Tail calls to functions with void returns
* Tail calls to abstract methods
* Tail calls to interface methods
* tail. calli sequences
* Tail calls to struct instance methods
* Tail calls involving refs
* Tail calls involving byrefs
* Tail calls involving generics

Remove some illegal tests that pass the address of variables on the local stack frame to a `tail.` call.

I have also made some changes to the tailcall return address hijacking test:
* On Linux x64, this test might not be forced to use helper because the struct `ForceHelper` is passed by value. I have added another arg to one of the tailcall helper functions that fixes this.
* I have changed the test to use fewer resources
* I am experimentally enabling it on ARM32

My new tests should fail on platforms without helper-based tailcalls and the hijacking one might also fail on ARM32/x64 Linux now. I will disable them once I see the CI results.

jakobbotsch commented:
It looks like the new test is running into #26311 on x64 Windows. I will have to disable it on all platforms except x86 Windows.

jakobbotsch commented:
It was strange to me that the issue in my previous comment was not more widespread since F* users should be running into it for many interface calls in tailcall position. I found out that this failure is a regression -- my new test works fine in .NET core 3.0 preview 8. After some bisecting I have tracked it down to #20643. I will leave it up to someone else whether we should try a targeted fix for it or wait for it to be fixed with #26418 considering that it is a regression.

cc @jashook

jakobbotsch commented:
Here is a smaller F# example that access violates with master on x64 Windows:
```fsharp
open System
open System.Runtime.CompilerServices

type IFoo =
    abstract member TailCallee : int -> Guid -> int

[<MethodImpl(MethodImplOptions.NoInlining)>]
let TailCaller (foo : IFoo) =
    foo.TailCallee 0 (new Guid())

type Bar() =
    interface IFoo with
        member this.TailCallee _ _ = 0

[<EntryPoint>]
let main argv = TailCaller (new Bar())

```

jashook commented:
@jakobbotsch this regression happens because we used to fast tail call, but we make the decision to not after #20643?

jakobbotsch commented:
@jashook Both are slow tail calls. If I had to take a wild guess it would be related to the resetting of the arg info here:
https://github.com/dotnet/coreclr/blob/f8f0c55b862ee4512c62b7eeb0fdae8a21af463a/src/jit/morph.cpp#L8309-L8316
But I have not at all tried to diagnose it, so it is a wild guess.

jakobbotsch commented:
This should be good to go @jkotas.

jkotas commented on tests/issues.targets:
```diff
@@ -194,10 +194,7 @@
-        </ExcludeList>
-        <ExcludeList Include="$(XunitTestBinBase)/JIT/Methodical/tailcall_v4/hijacking/*">
-            <Issue>13828</Issue>
+            <Issue>Uses illegal il tailcall pop ret</Issue>
```
I see all the invalid sequences in the test commented out with:

```
//            tail.      // tail.call, pop, ret sequence is never valid for .NET Core (but is accepted by .NET x64)
```

It would nice to delete the invalid tests so that we do not have to be excluding them forever.

jashook replied:
I think I would prefer to do that in a separate PR, where we remove all tests of a certain type. That way if we bring the feature back it is easier to undo. We have had issues with this in the past, ie native varargs.

jakobbotsch replied:
I can put it in a separate commit and then the PR can be rebased so I don't have to open another PR.

jakobbotsch replied:
It seems like these tests are actually already not being run, so I will just clean it up as well as I can.

jakobbotsch replied:
However there are other tests that are incompatible with many platforms. For instance, tests that implicit tailcalls are performed and a test that pinvokes kernel32. It does not seem we will be able to reenable this test any time soon. I guess I can remove those or try to condition them behind checks. What do you suggest?

jakobbotsch replied:
@jkotas the test tries a pinvoke to kernel32!Sleep. Is there a pinvokes declaration for some coreclr function I could use that would "just work" on all our supported platforms?

jkotas replied:
Can the test call `Thread.Sleep`?

jakobbotsch replied:
No, it is specifically testing `tail.` prefixed pinvoke call.

jakobbotsch replied:
Unless `Thread.Sleep` is just a public pinvoke?

jkotas replied:
For Unix, you can use `printf("Hello")` in `[DllImport("libc")]`

jkotas commented on tests/src/JIT/Directed/tailcall/more_tailcalls.il:
```diff
@@ -0,0 +1,3483 @@
+
+//  Microsoft (R) .NET Framework IL Disassembler.  Version 4.5.30319.0
```
Could you please replace this with the standard license header

jkotas commented on tests/src/JIT/Directed/tailcall/more_tailcalls.il:
```diff
@@ -0,0 +1,3483 @@
+
+//  Microsoft (R) .NET Framework IL Disassembler.  Version 4.5.30319.0
+
+
```
It is useful to keep the C# file that I assume you have used to generate this for reference. It is hard to understand and review what the pile of IL does otherwise.

jakobbotsch replied:
Should I just add it as a comment at the top? It is also relying on InlineIL.Fody, but I guess that is ok?

jkotas replied:
Comment at the top or separate file. Either way is fine.

jakobbotsch commented:
1. Added the .cs file for the new test
1. Made TailcallVerifyWithPrefix pinvoke `printf` on Unix and `Sleep` on Windows. Clarified that it is now disabled because of no helper-based tailcalls where it is disabled. I will reenable it on the new supported platforms in #26818.
1. Removed all the dead code from `TailcallVerifyWithPrefixTest`

jashook commented on tests/issues.targets:
```diff
@@ -891,6 +894,12 @@
+    <!-- Tests that run only on x86 Windows -->
+    <ItemGroup Condition="'$(XunitTestBinBase)' != '' and ('$(BuildArch)' != 'x86' or '$(TargetsWindows)' != 'true')">
+        <ExcludeList Include="$(XunitTestBinBase)/JIT/Directed/tailcall/more_tailcalls/*">
+            <Issue>x64 Windows has bugs with helper-based tailcall to VSD; rest of platforms do not support it</Issue>
```
Is there an open issue tracking this?

jashook replied:
If not, can you open one and add the issue number here?

jakobbotsch replied:
Will do, this issue is #26311. I believe it is the cause of a number of other issues opened recently, like #25960 and #26673.

jashook commented on tests/src/JIT/Directed/tailcall/more_tailcalls.cs:
```diff
@@ -0,0 +1,998 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using InlineIL;
```
Do you mind also adding a small readme into this folder explaining that this file is not used by testing, it is here to help regenerate the IL test.

jakobbotsch replied:
Added a comment to the top of the file

jashook commented:
Also, feel free to merge after the changes. This has passed CI. No reason to run them again.

jakobbotsch commented:
> Also, feel free to merge after the changes.

I don't have access to. :smile:
